lib/mongodb.js

```diff
@@ -2023,28 +2033,49 @@ function visitAllProperties(data, modelCtorOrDef, visitor) {
         visitAllProperties(value, def.type.definition, visitor);
       }
     } else {
-      visitor(value, def, (newValue) => { data[p] = newValue; });
+      visitor(model, value, def, (newValue) => { data[p] = newValue; });
     }
     continue;
   }
 }
 
 /**
- *
- * @param {*} propValue Property value to coerce into a Decimal128 value
- * @param {*} propDef Property definition to check if property is MongoDB
- * Decimal128 type
+ * @param {*} model Model containing the property
+ * @param {*} propValue Property value to coerce into special types supported by the connector
+ * @param {*} propDef Property definition to check if property is for MongoDB
  */
-function coerceDecimalProperty(propValue, propDef, setValue) {
-  let updatedValue;
-  if (hasDataType('decimal128', propDef)) {
-    if (Array.isArray(propValue)) {
-      updatedValue = propValue.map(val => Decimal128.fromString(val));
-      return setValue(updatedValue);
-    } else {
-      updatedValue = Decimal128.fromString(propValue);
-      return setValue(updatedValue);
+function coerceProperties(model, propValue, propDef, setValue) {
```

Comment:
I'd prefer having two distinct functions `coerceDecimalProperty` and `coerceObjectIdProperty`, but I understand that could be costly to call `visitAllProperties` twice.